onnxruntime/contrib_ops/cpu/reverse.cc

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "reverse.h"
+#include <utility>
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/framework/utils.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    Reverse,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllTensorTypes()),
+    Reverse);
+
+Status handle_scalar_tensor(const Tensor* input_tensor, Tensor* output_tensor, const MLDataType& dtype) {
+  // for scalar tensors, "reversing" is just copying the input buffer to the output buffer
+
+  // string scalar
+  if (dtype == DataTypeImpl::GetType<std::string>()) {
+    const std::string* src = input_tensor->template Data<std::string>();
+    std::string* dst = output_tensor->template MutableData<std::string>();
+    if (src != dst)
+      std::copy(src, src + 1, dst);
+    return Status::OK();
+  }
+
+  // MLFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<MLFloat16>()) {
+    const auto* src = input_tensor->Data<MLFloat16>();
+    auto* dst = output_tensor->template MutableData<MLFloat16>();
+    dst[0] = MLFloat16(math::floatToHalf(math::halfToFloat(src[0].val)));
+    return Status::OK();
+  }
+
+  // BFLoat16 scalar
+  if (dtype == DataTypeImpl::GetType<BFloat16>()) {
+    const auto* src = input_tensor->Data<BFloat16>();
+    auto* dst = output_tensor->template MutableData<BFloat16>();
+    dst[0] = BFloat16(src[0].ToFloat());
+    return Status::OK();
+  }
+
+  // non-string and non-float16 scalars - just copy the raw bytes
+  const void* src = input_tensor->DataRaw(dtype);
+  void* dst = output_tensor->MutableDataRaw(dtype);
+  if (src != dst)
+    memcpy(dst, src, 1 * dtype->Size());
+
+  return Status::OK();
+}
+
+template <int rank>
+void ReverseImplMLFloat16Type(const OpKernelContext* ctx, const Tensor* input_tensor, Tensor* output_tensor, const TensorShape& shape, const std::vector<int64_t>& reverse_axes) {
+  AllocatorPtr allocator;
+  ctx->GetTempSpaceAllocator(&allocator);
+  ORT_ENFORCE(ctx->GetTempSpaceAllocator(&allocator).IsOK(), "ORT needs couldn't get access to intermediate memory allocator to process float16 data in Reverse operator");
+  ORT_ENFORCE(allocator != nullptr, "ORT needs couldn't get access to intermediate memory allocator to process float16 data in Reverse operator");
+
+  const int64_t len = shape.Size();
+  ORT_ENFORCE(len > 0, "Need atleast one float16 element to be processed in Reverse operator");
+
+  // allocate intermediate buffers to be used for processing
+  float* input_buffer = static_cast<float*>(allocator->AllocArray(sizeof(float), len));
```

Comment:
Do we have any performance requirements for this operator? Feels like a lot of data movement to first convert to float, run eigen reverse, and then convert back to the data type.

Reply:
Well it's only really for float16 types hat we make the intermediate copy because we have to use Eigen. For all the other types (which are supposedly more common than float16), we use the raw buffers directly.